trieste/space.py

```diff
@@ -155,42 +155,58 @@ def __deepcopy__(self, memo: Dict[int, object]) -> DiscreteSearchSpace:
         return self
 
 
+Vector = TypeVar("Vector", tf.Tensor, List[float])
+""" A type variable representing either a :class:`tf.Tensor` or `list`. """
+
+
 class Box(SearchSpace):
     r"""
     Continuous :class:`SearchSpace` representing a :math:`D`-dimensional box in
     :math:`\mathbb{R}^D`. Mathematically it is equivalent to the Cartesian product of :math:`D`
     closed bounded intervals in :math:`\mathbb{R}`.
     """
 
-    def __init__(self, lower: TensorType, upper: TensorType):
-        """
-        :param lower: The lower (inclusive) bounds of the box.
-        :param upper: The upper (inclusive) bounds of the box.
-        :raise ValueError: If ``lower`` and ``upper`` have different shapes. Or if ``upper`` is not
+    def __init__(self, lower: Vector, upper: Vector):
+        r"""
+        If ``lower`` and ``upper`` are `list`\ s, they will be converted to tensors of dtype
+        `tf.float64`.
+
+        **Type hints:** If ``lower`` or ``upper`` is a `list`, they must both be `list`\ s.
+
+        :param lower: The lower (inclusive) bounds of the box. Must have shape [D] for positive D,
+            and if a tensor, must have float type.
+        :param upper: The upper (inclusive) bounds of the box. Must have shape [D] for positive D,
+            and if a tensor, must have float type.
+        :raise ValueError: If ``lower`` and ``upper`` have invalid shapes. Or if ``upper`` is not
             greater than ``lower`` across all dimensions.
         :raise TypeError: If ``lower`` and ``upper`` have different dtypes.
         """
+        tf.debugging.assert_shapes([(lower, ["D"]), (upper, ["D"])])
+        tf.assert_rank(lower, 1)
+        tf.assert_rank(upper, 1)
 
-        if not shapes_equal(lower, upper):
-            raise ValueError(
-                f"Lower and upper bounds must have the same shape, got {lower.shape} and"
-                f" {upper.shape}"
-            )
+        if len(lower) == 0:
+            raise ValueError(f"Bounds must have shape [D] for positive D, got {tf.shape(lower)}.")
 
-        if lower.dtype is not upper.dtype:
-            raise TypeError(
-                f"Lower and upper bounds must have the same dtype, got {lower.shape} and"
-                f" {upper.shape}"
-            )
+        if isinstance(lower, list):
+            self._lower = tf.cast(lower, dtype=tf.float64)
+            self._upper = tf.cast(upper, dtype=tf.float64)
+        else:
+            assert isinstance(lower, tf.Tensor) and isinstance(upper, tf.Tensor)  # to help mypy
```

Comment:
you could try `tf.convert_to_tensor(lower)` etc here, that way you'd support numpy arrays (it'll use the array dtype)

Reply:
musings:

not supporting numpy was semi-intentional, because I can see it being quite a maintenance overhead. However, `TensorType` include `ndarray`, so something would need to shift.

I like the idea of being able to provide `ndarray` support, I'm just wondering how one tests for `ndarray` support w/o duplicating every test. I'll think more about this

Reply:
this would probably less of a problem if tf and numpy had type hints. np will very soon, and tf in some unknown amount of time

Reply:
as for now, I've updated the code locally to support numpy. just looking at the tests